src/mono/mono/mini/interp/interp.c

```diff
@@ -1680,8 +1680,15 @@ ves_pinvoke_method (
 	}
 
 #ifdef MONO_ARCH_HAVE_INTERP_PINVOKE_TRAMP
+	gpointer call_info = *cache;
+
+	if (!call_info) {
+		call_info = mono_arch_get_interp_native_call_info (get_default_mem_manager (), sig);
+		mono_memory_barrier ();
+		*cache = call_info;
```

Comment:
We should use a CAS here to avoid leaks from races

Reply:
The memory is allocated from the mem manager, so its going to be leaked anyway.

Reply:
I see. Also a `mono_memory_write_barrier` here is enough.